### PR TITLE
TransposeConv with Bias

### PR DESCRIPTION
**Pull request by psunn**
### Description of issue:
For generating .tflite file with TFLiteConverter, when model contains
Conv2DTranspose layers, bias cannot fold into Operator TRANSPOSECONV.
It will result with extra Op ADD following Op TRANSPOSECONV.
But with other CONV-like layers (Conv2D, DepthwiseConv2D),
bias will be fold into CONV layer.
(check detailed TF issue: https://github.com/tensorflow/tensorflow/issues/34622)

### PR try to resolve it:
So we try to resolve this issue by enable TransposeConv with bias for TFLite:
  
- Update TFLite graph_transform features with:
      fill TransposeConv bias with zero if there is no bias
      fuse bias add into preceding TransposeConv(TEST added)

- Update TransposeConv with bias:
      add bias input to TransposeConv
      add optional bias to TransposeConv kernels

### example of the results:
  TRANSPOSE_CONV inputs:

1. output_shape
2. weights
3. activation
4. bias

fused_transposeconv

### Need to discuss:
~~currently this PR only update reference kernel for transpose_conv, optimised kernal is commented out.~~
~~several TEST need to be added as well, but~~ further suggestions are needed for adding additional test.

### CODE REVIEW

**Comment by jdduke** [comment]
Hi, thanks for the proposed CL. Until we have an optimized version of the fused bias, I'm not sure this is worth it. Whenever we change op semantics like this (e.g., adding a new attribute, or a new fused input), it impacts all delegates which support a given op. So we would also need to update our nnapi_delegate, gpu_delegate, etc... instances to handle the fused bias.

That may not be a bad thing, but I wonder how much that will really buy us in terms of performance? Have you measured the difference? transpose_conv isn't as common as those other operators, so extracting the absolutely maximal performance (like fusing the bias) isn't quite as critical.

**Comment by gbaned** [comment]
@psunn Could you please check jdduke's comments and keep us posted. Thanks!

**Comment by psunn** [comment]
@jdduke thanks for quick replying.
@gbaned thanks for remind me to reply.
I was in a week-long conference, sorry for late reply.

>>Until we have an optimized version of the fused bias, I'm not sure this is worth it.
I'll update code patch, as soon as I can.

>>That may not be a bad thing, but I wonder how much that will really buy us in terms of performance? Have you measured the difference?
I meet TransposeConv with Bias when I playing with [FSRCNN](https://arxiv.org/pdf/1608.00367.pdf), so I will try to measuring regarding this model.

**Comment by psunn** [comment]
> Until we have an optimized version of the fused bias, I'm not sure this is worth it.

I've updated the patch with optimized Ops. Please have a review.

> So we would also need to update our nnapi_delegate, gpu_delegate, etc... instances to handle the fused bias.

Looks like TransposeConv for nnapi [[1](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/nnapi/nnapi_delegate.cc#L2436)] and gpu_delegate [[1](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/gl/kernels/transpose_conv.cc#L83)] has bias inside, I think it will be a minor update.

**Comment by psunn** [comment]
> >So we would also need to update our nnapi_delegate, gpu_delegate, etc... instances to handle the fused bias.

>Looks like TransposeConv for nnapi [1] and gpu_delegate [1] has bias inside, I think it will be a minor update.

I think it's much easier from your side to push the changes regarding delegates, do you think so?

If it's must be pushed from us, shall we resolve it in a follow-up&separate PR, what do you think?

**Review by jdduke — CHANGES_REQUESTED** [review]
Can you add a test or two to transpose_conv_test.cc?

**Comment by psunn** [comment]
I've update the patch with new test, please have a review.

File: tensorflow/lite/kernels/transpose_conv.cc [hunk 1]
```diff
@@ -261,7 +266,20 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
   TF_LITE_ENSURE(context, input->type == kTfLiteFloat32 ||
                               input->type == kTfLiteUInt8 ||
                               input->type == kTfLiteInt8);
-  TF_LITE_ENSURE_EQ(context, weights->type, input->type);
+
+  if (hasBias) {
```
Comment (jdduke) [inline tensorflow/lite/kernels/transpose_conv.cc]: has_bias

File: tensorflow/lite/kernels/transpose_conv_test.cc [hunk 2]
```diff
@@ -63,6 +63,39 @@ class BaseTransposeConvOpModel : public SingleOpModel {
     }
     input_ = AddInput(input);
 
+    int bias_size = GetShape(filter_)[0];
+    if (input.type == TensorType_FLOAT32) {
+      bias_ = AddInput({TensorType_FLOAT32, {bias_size}});
```
Comment (jdduke) [inline tensorflow/lite/kernels/transpose_conv_test.cc]: We need to keep tests for the case that don't have a bias. Ideally we would keep the existing tests as is and just add a few additional ones that have biases.

**Review by jdduke — CHANGES_REQUESTED** [review]
OK, the last thing we need to do is update the op version when using a bias. To do that, you need to:

1) add a rule in lite/tools/versioning/op_version.cc to handle the TRANSPOSE_CONV with 4 inputs case and set the version to 3,
2) add an entry in lite/tools/versioning/op_version.cc for the new op version, and set it to kPendingReleaseOpVersion
3) Update lite/kernels/register.cc to use version 3 as the max_version for TRANSPOSE_CONV.

**Comment by psunn** [comment]
Thanks for review!

>     2\. add an entry in lite/tools/versioning/op_version.cc for the new op version, and set it to kPendingReleaseOpVersion

I guess you mean lite/toco/versioning/tflite/op_version.cc?

In lite/toco/versioning/tflite/op_version.cc, shall I add as:
```cpp
{{OperatorType::kTransposeConv, 1}, "1.9.0"},
{{OperatorType::kTransposeConv, 3}, kPendingReleaseOpVersion}
```
or  {{OperatorType::kTransposeConv, 2}, "x.x.x"} is also needed?

**Comment by jdduke** [comment]
Good catch, we should add `{{OperatorType::kTransposeConv, 2}, kPendingReleaseOpVersion}` as well, thanks!

**Comment by psunn** [comment]
I've update the patch, please have a review.

File: tensorflow/lite/kernels/transpose_conv_test.cc (hunk 2 above)
Comment (psunn) [reply tensorflow/lite/kernels/transpose_conv_test.cc]: done

**Comment by psunn** [comment]
@jdduke Can you have a review please?

**Comment by jdduke** [comment]
I'll be OOO for a few days, will review upon my return early next week (or I'll find somebody else to help review). Thanks again for your patience.

**Comment by psunn** [comment]
@jdduke sorry to bother you, I'm wondering is there any chance for some progress regarding this PR this week?

File: tensorflow/lite/toco/graph_transformations/fuse_binary_into_preceding_affine.cc [hunk 3]
```diff
@@ -27,6 +27,18 @@ namespace toco {
 
 namespace {
 
+int GetBiasIndex(const Operator& op) {
+  if (op.type == OperatorType::kConv ||
+      op.type == OperatorType::kFullyConnected ||
+      op.type == OperatorType::kDepthwiseConv) {
+        return 2;
```
Comment (jdduke) [inline tensorflow/lite/toco/graph_transformations/fuse_binary_into_preceding_affine.cc]: Nit: redunce indent

File: tensorflow/lite/toco/graph_transformations/fuse_binary_into_preceding_affine.cc [hunk 4]
```diff
@@ -27,6 +27,18 @@ namespace toco {
...
+  } else if (op.type == OperatorType::kTransposeConv) {
+        return 3;
```
Comment (jdduke) [inline tensorflow/lite/toco/graph_transformations/fuse_binary_into_preceding_affine.cc]: Same here, reduce indent

File: tensorflow/lite/toco/graph_transformations/ensure_bias_vectors.cc [hunk 5]
```diff
@@ -40,8 +41,19 @@ int GetOutputDepthFromWeights(const Model& model, const Operator& op) {
   return 0;
 }
 
+bool CheckOpInputSize(const Operator& op) {
+  if (op.type == OperatorType::kConv ||
+      op.type == OperatorType::kFullyConnected ||
+      op.type == OperatorType::kDepthwiseConv) {
+        return (op.inputs.size() >= 3);
```
Comment (jdduke) [inline tensorflow/lite/toco/graph_transformations/ensure_bias_vectors.cc]: Can you run the C++ formatter on this PR? https://www.tensorflow.org/community/contribute/code_style#c_coding_style

**Review by jdduke — COMMENTED** [review]
Looks good, just a couple nits.

File: tensorflow/lite/toco/graph_transformations/ensure_bias_vectors.cc (hunk 5 above)
Comment (psunn) [reply tensorflow/lite/toco/graph_transformations/ensure_bias_vectors.cc]: Thanks for review!
I've run them through clang-format, please have a check.

Comment (psunn) [reply tensorflow/lite/toco/graph_transformations/ensure_bias_vectors.cc]: I've rebased on master to resolve conflict.

**Comment by jdduke** [comment]
@suharshs can you validate the quantization-specific changes? Thanks.

**Comment by psunn** [comment]
I've rebased on master, so that pip-package can build without error.

**Comment by psunn** [comment]
@suharshs sorry to bother you, can you have a look this PR.

If nothing wrong, shall we try to merge it soon?

**Review by suharshs — DISMISSED** [review]
Thanks! This change looks fine, but note to get a working model with transpose conv biases, you will need to modify and test the post-training quantization tooling. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/operator_property.cc#L112

Feel free to tackle that in another PR.

**Comment by psunn** [comment]
Can we merge this? Will add test for post-training quantization tooling in separate PR.

**Review by jdduke — DISMISSED** [review]

**Comment by psunn** [comment]
Hi, any chance to merge this PR?

**Comment by jdduke** [comment]
Sorry for the delay, there are some internal conflicts that we have to resolve manually, hoping to resolve in the next day or two.

**Comment by psunn** [comment]
Hi @jdduke @suharshs , can you please review and hopefully re-approve please. I rebased this PR on latest master and resolved conflict. Thanks.

**Comment by suharshs** [comment]
Hi peng, this is messing with some internal tests that I am cleaning up.
Should be merged once that is done. Thanks!

**Comment by psunn** [comment]
Hi @jdduke @suharshs ,
Is it ok to re-approval this PR?
Hopefully merged soon, if no issues.

**Comment by suharshs** [comment]
I am still working through some internal tests that are failing. Will merge
the pr once those are resolved.

**Comment by suharshs** [comment]
Just an update here: there is an issue i am running into with flatbuffer import in the mlir code path. I am trying to get that fixed, but the optional tensor support has trickled into a few things, will update here when that is resolved. thanks for your patience.

File: tensorflow/lite/toco/graph_transformations/fuse_binary_into_preceding_affine.cc [hunk 6]
```diff
@@ -253,14 +268,22 @@ ::tensorflow::Status FuseBinaryIntoPrecedingAffine::Run(Model* model,
...
+      preceding_op->type != OperatorType::kDepthwiseConv &&
+      preceding_op->type != OperatorType::kTransposeConv) {
     AddMessageF(
         "Not fusing %s because the preceding %s is not of one of the supported "
         "types",
         LogName(*binary_op), LogName(*preceding_op));
     return ::tensorflow::Status::OK();
   }
 
+  if (preceding_op->type == OperatorType::kTransposeConv &&
```
Comment (suharshs) [inline tensorflow/lite/toco/graph_transformations/fuse_binary_into_preceding_affine.cc:280]: quick question, why did you have to add this if statement?

 i am still tackling some broken tests from an internal team and this is related.

Comment (psunn) [reply tensorflow/lite/toco/graph_transformations/fuse_binary_into_preceding_affine.cc:280]: I was intend to fused kAdd to kTransposeConv.
Remove if statement, kMul, kSub, kDiv may also be fused to kTransposeConv, which is also good for graph optimise.

Shall I remove it?

**Comment by suharshs** [comment]
No that's ok, was just wondering. Thanks

**Comment by gbaned** [comment]
@psunn Could you please resolve the conflicts? Thanks!

**Comment by psunn** [comment]
> @psunn Could you please resolve the conflicts? Thanks!

@gbaned done, thanks for the reminder.

**Comment by rthadur** [comment]
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.

**Pull request closed without merging**